tesseract_collision/core/src/types.cpp

```diff
@@ -57,57 +57,209 @@ void ContactResult::clear()
+                   (r.cc_type[j] == tesseract_collision::ContinuousCollisionType::CCType_Time1 || discrete))
+            r.cc_type[j] = tesseract_collision::ContinuousCollisionType::CCType_Time1;
+          else
+            r.cc_type[j] = tesseract_collision::ContinuousCollisionType::CCType_Between;
+
+          // If discrete set cc_transform for discrete continuous
+          if (discrete)
+            r.cc_transform = r.transform;
+        }
+      }
+    }
+
+    // Filter results
+    if (filter != nullptr)
+      filter(pair);
+
+    if (!pair.second.empty())
+    {
+      // Add results to the full segment results
+      cnt_ += static_cast<long>(pair.second.size());
+      auto it = data_.find(pair.first);
+      if (it == data_.end())
+      {
+        data_.insert(pair);
+      }
+      else
+      {
+        assert(it != data_.end());
+        // Note: Must include all contacts throughout the trajectory so the optimizer has all the information
+        //      to understand how to adjust the start and end state to move it out of collision. Originally tried
+        //      keeping the worst case only but ran into edge cases where this does not work in the units tests.
 
-  return v.size();
+        it->second.reserve(it->second.size() + pair.second.size());
+        it->second.insert(it->second.end(), pair.second.begin(), pair.second.end());
+      }
+    }
+  }
 }
 
-std::size_t flattenCopyResults(const ContactResultMap& m, ContactResultVector& v)
+long ContactResultMap::count() const { return cnt_; }
+
+std::size_t ContactResultMap::size() const
 {
-  v.clear();
-  v.reserve(m.size());
-  for (const auto& mv : m)
+  if (cnt_ == 0)
+    return 0;
+
+  std::size_t cnt{ 0 };
+  for (const auto& pair : data_)
   {
-    v.reserve(v.size() + mv.second.size());
-    std::copy(mv.second.begin(), mv.second.end(), std::back_inserter(v));
+    if (!pair.second.empty())
+      ++cnt;
   }
 
-  return v.size();
+  return cnt;
 }
 
-std::size_t flattenWrapperResults(ContactResultMap& m, std::vector<std::reference_wrapper<ContactResult>>& v)
+bool ContactResultMap::empty() const { return (cnt_ == 0); }
+
+void ContactResultMap::clear()
+{
+  if (cnt_ == 0)
+    return;
+
+  // Only clear the vectors so the capacity stays the same
+  for (auto& cv : data_)
+    cv.second.clear();
+
+  cnt_ = 0;
+}
+
+void ContactResultMap::release()
+{
+  data_.clear();
+  cnt_ = 0;
+}
+
+const ContactResultMap::ContainerType& ContactResultMap::getContainer() const { return data_; }
+
+ContactResultMap::ConstIteratorType ContactResultMap::begin() const { return data_.begin(); }
+
+ContactResultMap::ConstIteratorType ContactResultMap::end() const { return data_.end(); }
+
+ContactResultMap::ConstIteratorType ContactResultMap::cbegin() const { return data_.cbegin(); }
+
+ContactResultMap::ConstIteratorType ContactResultMap::cend() const { return data_.cend(); }
+
+const ContactResultVector& ContactResultMap::at(const KeyType& key) const { return data_.at(key); }
```

Comment:
Do these combined with clear result in any non intuitive behavior? For example, a range based for loop will iterate over things after you call clear. That's probably fine. I was trying to come up with some way that this could internally store another iterable structure that is a view into the actual data store, but that's probably not worth it.

Reply:
Yea, I think a better approach would be to leverage a object pool for ContactResultVector then have everything else function as a normal map and update the set and add methods to pull from the object pool if the pair does not exist. I have found some examples and will most likely implement this over the next few weeks.